### ISSUE TITLE
2.13: move to new collections

### ISSUE THREAD

**Issue by SethTisue** [scala/community-builds#710]
774c22c58973ee095d76cda5fc188255e742854c flips the switch. now to deal with the failures

**Comment by SethTisue** [comment]
after 1338c90cbd98c31d219e7f6d1b8f78496c52902e, in local testing,

* these are green (10): scala-xml, cloc-plugin, machinist, sbt-testng, scala-collection-compat, scala-js-stubs, scala-partest, scala-partest-interface, scalaj-http, sourcecode
* these key projects fail but I suggest we try hard to get them green before M4 (4): macro-paradise, scala-parser-combinators, scalacheck, kind-projector
* projects currently w/ upstream failures, but also probably essential for M4 (4): scalatest, scala-java8-compat, scala-parallel-collections, mima
* stretch goals for M4 (4): akka-actor, scalaz, specs2, scala-js
* I had to work around a scala-xml issue that will go away once scala-xml is out of the bootstrap; I will prioritize that PR

**Comment by lrytz** [comment]
for scalacheck there's a pr / branch. on my phone right now

**Comment by lrytz** [comment]
https://github.com/rickynils/scalacheck/pull/391 / https://github.com/lrytz/scalacheck/tree/newCollections?files=1

**Comment by SethTisue** [comment]
@lrytz thanks, using that now. and I fixed kind-projector.

nscala-time and sbinary are now giving:

```
[nscala-time] [error] **** Missing dependency: the library org.scalacheck#scalacheck is not provided (in space "scala") by any project in this configuration file.
```

which puzzles me. perhaps `set publishArtifact in packageDoc in jvm := false` is somehow not as selective as it looks

**Comment by lrytz** [comment]
Oh I forgot to mention: if you build scalacheck from my branch (https://github.com/lrytz/scalacheck/tree/newCollections), the groupId changed to org.scala-lang.modules (https://github.com/lrytz/scalacheck/commit/2013eca70ca1776d455e9ec50cd7716e4098c77d). So you'd have to revert that.

**Comment by SethTisue** [comment]
> the groupId changed to org.scala-lang.modules

ahhhhhhhhh :-)

**Comment by xuwei-k** [comment]
https://github.com/scalaz/scalaz/pull/1730

**Comment by julienrf** [comment]
https://github.com/scala/scala-parser-combinators/pull/148

**Comment by xuwei-k** [comment]
fixed scalaprops https://github.com/scalaprops/scalaprops/commit/b98d71cc43178cd2be9122bf956d1377c48fea25

**Comment by lrytz** [comment]
For the record, to not duplicate work: I'm working on scala-java8-compat (almost done), i have a local path for genjavadoc, and @ennru created a branch for akka-actor (https://github.com/ennru/akka/tree/newCollections).

**Comment by julienrf** [comment]
@SethTisue kind-projector compiles without warning and the tests pass with scala-2.13.0-pre-1c0be29. Can you point me to the error you faced?

**Comment by SethTisue** [comment]
@julienrf it was https://github.com/non/kind-projector/pull/68

**Comment by adriaanm** [comment]
I'll take macro-paradise

**Comment by milessabin** [comment]
@adriaanm ping me when you're done with that and I'll get shapeless into shape.

**Comment by SethTisue** [comment]
status report:

* green key projects: kind-projector, scala-collection-compat, scala-js-stubs, scala-parser-combinators, scala-xml, scalacheck, scalaz
* also green: cloc-plugin, machinist, nscala-time, sbt-testng, scala-partest, scalaj-http, sourcecode
* **key projects failing, high priority for M4: scalatest, macro-paradise**
* stretch goals for M4: scala-java8-compat, scala-parallel-collections, genjavadoc, akka-actor, specs2, scala-js, shapeless
* more stretch goals for M4: sbinary, fastparse, scalaprops

we would like to get M4 out in time for Scala Days Berlin (May 14)

**Comment by SethTisue** [comment]
macro-paradise progress can be followed at https://github.com/scala/scala/pull/6606

**Comment by SethTisue** [comment]
these are the currently failing projects. as in all their upstream dependencies are green, but the project itself fails:

* scala-refactoring, scalaprops, tut, sbinary, scala-js, sbt-io, scala-async, json4s, utest, scalatest, scala-collections-laws, scala-continuations

volunteers wanted, to tackle particular projects. or at least to investigate and report back on the nature of the failure and the expected difficulty of fixing it

https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1091/consoleFull has the failure logs.

**Comment by SethTisue** [comment]
@xuwei-k there is now a further error in sbinary:

```
[sbinary] [error] /Users/tisue/community.213/target-0.9.12/project-builds/sbinary-a0f53c0c59caf8e1feefc3b1a5e88ad36a208588/core/target/scala-2.13.0-pre-ac5c8fa/src_managed/main/standardtypes.scala:1055: not enough arguments for method apply: (from: Nothing)scala.collection.mutable.Builder[T,CC[T]] in trait BuildFrom.
[sbinary] [error] Unspecified value parameter from.
[sbinary] [error]         val builder = cbf.apply()
[sbinary] [error]                                ^
[sbinary] [error] /Users/tisue/community.213/target-0.9.12/project-builds/sbinary-a0f53c0c59caf8e1feefc3b1a5e88ad36a208588/core/target/scala-2.13.0-pre-ac5c8fa/src_managed/main/standardtypes.scala:1055: not enough arguments for method apply: (from: Nothing)scala.collection.mutable.Builder[T,CC[T]] in trait BuildFrom.
[sbinary] [error] Unspecified value parameter from.
[sbinary] [error]         val builder = cbf.apply()
[sbinary] [error]                                ^
```

**Comment by xuwei-k** [comment]
- https://github.com/propensive/magnolia/pull/93
- https://github.com/lihaoyi/utest/pull/163

**Comment by xuwei-k** [comment]
fixed some compile errors in sbt/io 1.1.x branch https://github.com/sbt/io/commit/6c4364e6348e3f34c46eefac3db6596d3ee5cf54 but community-builds reference v1.1.6  https://github.com/scala/community-builds/blob/dd1ed340ea8718156d4bf943532c00e5fa9a9a2e/configs/project-refs.conf#L98

**Comment by xuwei-k** [comment]
https://github.com/kxbmap/configs/pull/30

**Comment by xuwei-k** [comment]
https://github.com/json4s/json4s/commit/4053ceedde625524cbfaf34d155fbbc6c1013236

**Comment by xuwei-k** [comment]
https://github.com/kxbmap/configs/pull/31

**Comment by xuwei-k** [comment]
https://github.com/scalikejdbc/scalikejdbc/pull/887

**Comment by xuwei-k** [comment]
https://github.com/scallop/scallop/pull/166

**Comment by milessabin** [comment]
https://github.com/milessabin/macro-compat/commit/b1bbe40105a649abf35511fe8e7de4c0419192f7

**Comment by milessabin** [comment]
https://github.com/milessabin/shapeless/tree/shapeless-2.3.3-scala-2.13.0-M4

**Comment by xuwei-k** [comment]
https://github.com/argonaut-io/argonaut/pull/304

**Comment by lrytz** [comment]
scalacheck 1.14.0 for M4 is now available (http://search.maven.org/#artifactdetails%7Corg.scalacheck%7Cscalacheck_2.13.0-M4%7C1.14.0%7Cjar)

**Comment by MasseGuillaume** [comment]
@SethTisue I'm doing akka-actor. It looks like there is prior work here: https://github.com/ennru/akka/commit/a89c4faa73f0ff6d7ef854a7315bd757865f7a6f

**Comment by SethTisue** [comment]
@MasseGuillaume 👍. I'd suggest opening a ticket in akka/akka to coordinate with the Akka team about it

**Comment by SethTisue** [comment]
@xuwei-k thanks, your changes have made some projects green; others are waiting on scalatest and/or specs2. 

what do you make of the argonaut errors at https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1169/consoleFull?

**Comment by SethTisue** [comment]
overall status currently is: I won't be giving this much attention until after Scala Days NYC, but hopefully by then ScalaTest and specs2 will have done their M4 updates. that will unblock a ton of projects here.

the current failing list (from this [example recent run](https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1169/console)):

* scodec-bits, scala-js, jsoniter-scala, magnolia, scala-continuations, scala-async, scala-collections-laws, argonaut, utest, scala-refactoring, singleton-ops, sbinary, tut, scalatest, specs2

**Comment by SethTisue** [comment]
oh and there are 20 green projects currently: cloc-plugin, json4s, kind-projector, kxbmap-configs, machinist, macro-compat, minitest, nscala-time, sbt-testng, scala-collection-compat, scala-js-stubs, scala-parser-combinators, scala-xml, scalacheck, scalaj-http, scalalib, scalaprops, scalaz, shapeless, sourcecode

**Comment by xuwei-k** [comment]
> what do you make of the argonaut errors at https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1169/consoleFull?

```
[argonaut] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.12/project-builds/argonaut-ed917ca0da1db65fc3a44c03c13288bd77f55de5/argonaut/jvm/src/test/scala/argonaut/ArgonautSpec.scala:3: object specs2 is not a member of package org
[argonaut] [error] import org.specs2._
[argonaut] [error]            ^
[argonaut] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.12/project-builds/argonaut-ed917ca0da1db65fc3a44c03c13288bd77f55de5/argonaut/jvm/src/test/scala/argonaut/ArgonautSpec.scala:5: not found: type Specification
[argonaut] [error] trait ArgonautSpec extends Specification with ScalaCheck {
[argonaut] [error]                            ^
```

argonaut disabled specs2 and scalacheck dependency for compile only check in travis-ci 😅 

- https://github.com/argonaut-io/argonaut/blob/59a81e7e64af19bdc57cb52d3a63f3595668b701/project/build.scala#L130-L139
- https://github.com/argonaut-io/argonaut/commit/96ce4b616e1d09
- https://github.com/argonaut-io/argonaut/blob/59a81e7e64af19bdc57cb52d3a63f3595668b701/.travis.yml#L31-L32

**Comment by xuwei-k** [comment]
https://github.com/sbt/sbinary/pull/16

**Comment by soronpo** [comment]
singleton-ops should now be in the green thanks to this PR by @xuwei-k 
https://github.com/fthomas/singleton-ops/pull/68

**Comment by lrytz** [comment]
`scala-java8-compat` 0.9.0 is released. In this case we went for a separate branch, because the project exploits lots of internals of the collection library, and cross-building would be too cumbersome (https://github.com/scala/scala-java8-compat/commits/2.13.x).

**Comment by MasseGuillaume** [comment]
I made a graph of the community build with scaladex. The bottom part is the critical path: scala-parser-combinators, scala-xml, macro-paradise, etc.

moved to: https://github.com/scala/community-builds/wiki/scala-2.13.x-migration#whats-next-

It's also possible to make similar graphs not limited to the community build.

**Comment by xuwei-k** [comment]
https://github.com/sirthias/parboiled/pull/120

**Comment by MasseGuillaume** [comment]
moved to: https://github.com/scala/community-builds/wiki/scala-2.13.x-migration

**Comment by xuwei-k** [comment]
- https://github.com/eed3si9n/sjson-new/pull/95
- https://github.com/lightbend/scala-logging/pull/124
- https://github.com/unfiltered/unfiltered/commit/6150fc42271ec45b1f43e5d715f963

**Comment by xuwei-k** [comment]
https://github.com/scalapb/protobuf-scala-runtime/pull/21

**Comment by dwijnand** [comment]
Should we create a wiki page, in either this repo or scala/scala, to track this?

(ala https://github.com/sbt/sbt/wiki/sbt-1.x-plugin-migration / https://github.com/sbt/sbt/wiki/library-sbt-1.x-migration)

**Comment by MasseGuillaume** [comment]
@dwijnand https://github.com/scala/community-builds/wiki/scala-2.13.x-migration

**Comment by SethTisue** [comment]
currently (here's a [recent run](https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1240/console)) 41 projects are green

the highest priority failures (the ones blocking the most projects downstream are):

* akka-actor (the Akka team is working on it, with Guillaume M)
* utest (needs volunteer)
* minitest (Alexandru is on vacation)
* cats (Typelevel folks are working on publishing for M4, but could perhaps use volunteer assistance)
* shapeless (@milessabin?)

the full failures list (not including projects blocked by failing dependencies) currently is: magnolia, scala-js, scala-refactoring, scala-async, minitest, shapeless, scala-continuations, utest, scala-collections-laws, linter, spray-json, twirl, scallop, sksamuel-exts, scalamock, paradox, mima, scalajson, grizzled, slick, scoverage, jackson-module-scala, scala-stm, parboiled, akka-actor, scapegoat, pcplod, scalikejdbc, lift-json, cachecontrol, scalastyle, scalaz8, log4s, http4s-parboiled2, scalatest-tests, scala-swing, cats, tut

misc notes:
* scala-js and spray-json are blocked by scala-parallel-collections being missing
* I just pushed a commit that will hopefully make jackson-module-scala green

**Comment by dwijnand** [comment]
In addition to https://github.com/scala/community-builds/wiki/scala-2.13.x-migration
see also https://github.com/scala/make-release-notes/blob/2.13.x/projects-2.13.md

**Comment by sjrd** [comment]
For Scala.js, while waiting for the parallel collections, you could try pruning [this line](https://github.com/scala/community-builds/blob/51704def2869eab5799e05fac0def54a806a9d82/configs/community.dbuild#L357) to only keep `"testSuite"`, removing `io`, `logging` and `linker`. The `testSuite` is supposed to pass.

**Comment by milessabin** [comment]
@SethTisue scala/scala PR addressing the shapeless failure here: https://github.com/scala/scala/pull/6971.

**Comment by xuwei-k** [comment]
Create new wiki page for 2.13.0-M5? or reuse https://github.com/scala/community-builds/wiki/scala-2.13.x-migration?

**Comment by dwijnand** [comment]
I'd say let's keep that URL from M5, and move the M4 content to another page.

**Comment by sjrd** [comment]
utest should be back on track https://github.com/lihaoyi/utest/pull/180

**Comment by SethTisue** [comment]
Haven't posted an update here in a while, since I have been concentrating instead on encouraging library authors to publish for 2.13.0-M5, since usually that results in the project going green in the community build as well.

but anyway, the current status in a recent run such as https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1529/ is:

```
unexpected FAILED: ssl-config
SUCCESS 64 FAILED 41 DID NOT RUN 83 TOTAL 188
UNEXPECTED 1
BLOCKERS:
42 cats
22 scala-java8-compat
22 fastparse
20 ssl-config
15 geny
14 pprint
14 scalatags
10 jackson-module-scala
9 jawn-0-10
6 twirl
5 scala-async
5 cachecontrol
3 slick
3 scalamock
3 tut
2 log4s
2 pcplod
2 grizzled
1 http4s-parboiled2
1 base64
1 boopickle
1 perfolation
1 lift-json 
```

once cats is published for M5, it should go green here as well. scala-java8-compat is on my plate, but it would be awesome if someone volunteered to tackle it. ssl-config can wait until #808 is merged. almost everything else on the list is probably fair game if anyone wants to try to help the maintainers move to 2.13.

**Comment by SethTisue** [comment]
(note that the logs on Jenkins now end with these stats and the sorted list, I added that in August, 1b5e0e6869fbf67e0b7a8f83c8281f5c57489ada)

**Comment by SethTisue** [comment]
a sample recent run is https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1593/console

the big news is that cats is now green... except it's missing cats-kernel-laws, which several downstream projects need (algebra, circe, kittens, paiges)

the current top-blockers list is:

```
21 scala-java8-compat   (@lrytz is looking into it)
19 fastparse
12 geny
12 scalatags
11 cats-effect
11 pprint
11 paiges
10 jackson-module-scala
9 jawn-0-10
8 circe
5 scala-async
5 cachecontrol
3 slick
3 scalamock
3 algebra
```

the numbers are how many downstream projects are blocked

**Comment by SethTisue** [comment]
* geny: 4e4715eb76635173b78cd99a14ca4aa600b8e28c
* fastparse: 830c7c2c28ab764ec6c891cc21f17ef2720fd86d
* cats-kernel-laws: 7d058154eb27f88d8790a9ddf58f2bd9aa95abb0
* cats-effect: 7249547b3f76e709d24f57cb5c81036365c02c6d

let's see how https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1625/ does

**Comment by SethTisue** [comment]
at https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/1630/ we're up to 78 green projects

```
unexpected FAILED: fastparse
SUCCESS 78 FAILED 47 DID NOT RUN 63 TOTAL 188
UNEXPECTED 1
BLOCKERS:
21 scala-java8-compat
19 fastparse
11 pprint
10 jackson-module-scala
9 jawn-0-10
5 scala-async
5 cachecontrol
3 slick
3 scalamock
2 specs2-more
2 log4s
2 jsoniter-scala
2 monix
2 fs2
2 spire
2 pcplod
2 tut
2 grizzled
```

**Comment by SethTisue** [comment]
```
[fastparse] [error] /home/jenkins/workspace/scala-2.13.x-integrate-community-build/target-0.9.16/project-builds/fastparse-e5427c7be70c56d55f818bd736a6cce79443c8bc/scalaparse/shared/src/main/scala/scalaparse/Core.scala:83:17: not found: value _
[fastparse] [error]   val `_*` = P( `_` ~ `*` )
[fastparse] [error]                 ^
```

looks like fastparse ran afoul of https://github.com/scala/scala/pull/6218, @som-snytt does this look as expected?

**Comment by som-snytt** [comment]
I'm taking a look, but building fastparse against my local scalac snapshot is non-trivial because of dependencies. Is there a quick way for me to compile fastparse against some magic repo at lightbend?

I was about to just use scalac, which I don't mind for a one-off test, but `SequencerGen` is generated in the build, so everything is quickly out of scope.

The linked PR just specifies how to import a member named underscore. It should just work, unless a macro is fumbling an import.

**Comment by SethTisue** [comment]
> building fastparse against my local scalac snapshot is non-trivial because of dependencies. Is there a quick way for me to compile fastparse against some magic repo at lightbend

no (https://github.com/scala/community-builds/issues/611)

the easiest thing is to use dbuild with a reduced config file so it doesn't extract dependencies from 180 projects you don't care about. I made this branch of the community build that has only fastparse and its dependencies: https://github.com/SethTisue/community-builds/tree/fastparse-only . just grab the branch and `./run.sh`

**Comment by soronpo** [comment]
>  I made this branch of the community build that has only fastparse and its dependencies

Perhaps a nice feature will be to able to specify the required library as an (optional) argument and the community build script will only build that specific library and its dependencies.

**Comment by SethTisue** [comment]
> Perhaps a nice feature will be to able to specify the required library as an (optional) argument and the community build script will only build that specific library and its dependencies.

note that it does something like that already, in the sense that `./run.sh fastparse` will build only fastparse and its dependencies. but in order to discover what those dependencies even are, every project must be extracted, which takes a while. the changes @cunei is working on at https://github.com/lightbend/dbuild/issues/210 will speed up followup runs, but not first runs. so the "delete all the projects you don't care about" trick will remain useful even after Toni's changes land

so, yes: it would be nice if there was a way to say "don't even extract any projects I didn't mention on the command line". but deleting the entries you don't care about from the config file is generally not that hard, so it's not absolutely clear it would be worth Toni's time to automate it more, given that dbuild only has a few users

**Comment by soronpo** [comment]
Is it possible to somehow query scaladex to achieve this easier?

**Comment by SethTisue** [comment]
@soronpo scaladex only has runtime dependencies. in this context we need compile-time dependencies, including test-only dependencies.

**Comment by som-snytt** [comment]
Another great feature would be if Seth could stop by my cubicle. I ran the special branch but got errors, I'll spend time with it later.

**Comment by SethTisue** [comment]
I've deleted 
https://github.com/scala/community-builds/wiki/scala-2.13.x-migration as it had fallen completely out of date. sorry, I know this was well-intentioned, but I didn't find it useful in my own (informal) process.

the best way for anyone to check the current status is to look at the log of a recent run at https://scala-ci.typesafe.com/job/scala-2.13.x-integrate-community-build/ and the expected-to-fail project list at https://github.com/scala/community-builds/blob/2.13.x/report/Report.scala (which I do actively maintain)

**Comment by SethTisue** [comment]
overall the current status is: we've been hovering at around 80 green projects. I've been tackling any significant regressions myself, enlisting project maintainers' help as needed

it's now been so long since M5, and so many little breaking changes have accumulated, that it's not really productive at present to do any better than ~80

after 2.13.0-RC1 is out, we can make a fresh attempt to get more projects green.

> I have been concentrating instead on encouraging library authors to publish for 2.13.0-M5, since usually that results in the project going green in the community build as well.

(meanwhile, I have continued this nagging campaign)

**Comment by dwijnand** [comment]
I think https://github.com/scala/community-builds/wiki/scala-2.13.x-migration served the initial purpose of bootstrapping the ecosystem for the new breaking pre-release of 2.13.  When 2.13.0-RC1 comes out it'll be useful to swarm over those usual suspect libraries that'll need updating.  And when 2.13.0 is out we'll port even more than usual.

**Comment by SethTisue** [comment]
potential 2.13.0-RC1 release blockers in the community build are being tracked at https://github.com/scala/bug/issues/11453 until RC1 is out

**Comment by SethTisue** [comment]
RC1 is out. I'm going to close this ticket, since:

* enough of the build (94 projects) was green to permit RC1 to come out
* now that RC1 is out, most of the action now will happen because maintainers themselves will publish for RC1 and 2.13.0, and their projects will go green in the community build as a result
* improving the 2.13 community build is a constant background task for me, we don't need an endlessly open ticket on it

**Comment by SethTisue** [comment]
RC2 is out

122 projects are green now (up from 94 six weeks ago)

there aren't currently isn't anything readily actionable blocking many downstream projects. `BLOCKERS: 12 scalameta, 9 jawn-0-10, 3 twitter-util, 3 scalamock, 3 algebra, 3 mockito-scala, 3 tut, 2 enumeratum, 1 doobie, 1 lift-json, 1 monix, 1 elastic4s, 1 circe-jackson, 1 sttp, 1 magnolia, 1 parboiled2, 1 tsec, 1 classutil`

scalameta (blocking 12) is difficult, see #910

jawn 0.10 (blocking 9) isn't worth upgrading, we only need it for building sbt and its modules. we can wait for sbt itself to move to 2.13

and the other blocker numbers are small. I'll chip away it in the coming months. (community help with this is welcome)

**Comment by SethTisue** [comment]
162 green projects these days (2.12 has 204)
